rl/agents/sarsa.py

```diff
@@ -0,0 +1,471 @@
+
+        self.target_model = clone_model(self.model, self.custom_model_objects)
+        self.target_model.compile(optimizer='sgd', loss='mse')
+        self.model.compile(optimizer='sgd', loss='mse')
+
+        if self.target_model_update < 1.:
+            # We use the `AdditionalUpdatesOptimizer` to efficiently soft-update the target model.
+            updates = get_soft_target_model_updates(self.target_model, self.model, self.target_model_update)
+            optimizer = AdditionalUpdatesOptimizer(optimizer, updates)
+
+        """
+
+        def clipped_masked_error(args):
+            y_true, y_pred, mask = args
+            loss = huber_loss(y_true, y_pred, self.delta_clip)
+            loss *= mask  # apply element-wise mask
+            return K.sum(loss, axis=-1)
+
+        # Create trainable model. The problem is that we need to mask the output since we only
+        # ever want to update the Q values for a certain action. The way we achieve this is by
+        # using a custom Lambda layer that computes the loss. This gives us the necessary flexibility
+        # to mask out certain parameters by passing in multiple inputs to the Lambda layer.
+        y_pred = self.model.output
+        y_true = Input(name='y_true', shape=(self.nb_actions,))
+        mask = Input(name='mask', shape=(self.nb_actions,))
+        loss_out = Lambda(clipped_masked_error, output_shape=(1,), name='loss')([y_pred, y_true, mask])
+        ins = [self.model.input] if type(self.model.input) is not list else self.model.input
+        trainable_model = Model(input=ins + [y_true, mask], output=[loss_out, y_pred])
+        assert len(trainable_model.output_names) == 2
+        combined_metrics = {trainable_model.output_names[1]: metrics}
+        losses = [
+            lambda y_true, y_pred: y_pred,  # loss is computed in Lambda layer
+            lambda y_true, y_pred: K.zeros_like(y_pred),  # we only include this for the metrics
+        ]
+        trainable_model.compile(optimizer=optimizer, loss=losses, metrics=combined_metrics)
+        self.trainable_model = trainable_model
+
+        self.compiled = True
+
+    def load_weights(self, filepath):
+        self.model.load_weights(filepath)
+        """
+        # there is no need of target model in SARSA Algorithm
+        self.update_target_model_hard()
+        """
+
+    def save_weights(self, filepath, overwrite=False):
+        self.model.save_weights(filepath, overwrite=overwrite)
+
+    def reset_states(self):
+        self.recent_action = None
+        self.recent_observation = None
+        if self.compiled:
+            self.model.reset_states()
+
+            """
+            # there is no need of target model in SARSA Algorithm
+            self.target_model.reset_states()
+            """
+
+    def update_target_model_hard(self):
+        assert False, 'target model is disabled in SARSA'
+
+    def fit(self, env, nb_steps, action_repetition=1, callbacks=None, verbose=1,
```

Comment:
What about the following:

- have two variables, `state0` and `state1`.
- Whenever `forward()` is called, set `state0 = state1` and `state1 = observation`.
- Then you have access to both subsequent states in `backward()` without the need to update anything. This would essentially delay the update by a single timestep but I don't think that is going to be a problem.

Would that work?